server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -130,47 +135,73 @@ protected void doExecute(Task task, FieldCapabilitiesRequest request, final Acti
             indexResponses.putIfAbsent(resp.getIndexName(), resp);
         };
         final FailureCollector indexFailures = new FailureCollector();
-        // One for each cluster including the local cluster
-        final CountDown completionCounter = new CountDown(1 + remoteClusterIndices.size());
-        final Runnable countDown = createResponseMerger(request, fieldCapTask, completionCounter, indexResponses, indexFailures, listener);
-        final RequestDispatcher requestDispatcher = new RequestDispatcher(
-            clusterService,
-            transportService,
-            task,
-            request,
-            localIndices,
-            nowInMillis,
-            concreteIndices,
-            threadPool.executor(ThreadPool.Names.SEARCH_COORDINATION),
-            handleIndexResponse,
-            indexFailures::collect,
-            countDown
-        );
-        requestDispatcher.execute();
+        try (RefCountingRunnable refs = new RefCountingRunnable(() -> {
+            indexMappingHashToResponses.clear();
+            if (fieldCapTask.notifyIfCancelled(listener)) {
+                LOGGER.trace("clear index responses on cancelled");
+                indexResponses.clear();
+                indexFailures.clear();
+            } else {
+                mergeIndexResponses(request, fieldCapTask, indexResponses, indexFailures, listener);
+            }
+        })) {
+            // local cluster
+            final var localRef = Releasables.releaseOnce(refs.acquire());
+            fieldCapTask.addListener(() -> fieldCapTask.notifyIfCancelled(ActionListener.releasing(localRef)));
+            final RequestDispatcher requestDispatcher = new RequestDispatcher(
+                clusterService,
+                transportService,
+                task,
+                request,
+                localIndices,
+                nowInMillis,
+                concreteIndices,
+                threadPool.executor(ThreadPool.Names.SEARCH_COORDINATION),
+                indexResponse -> {
+                    try (Releasable ref = refs.tryAcquire()) {
```

Comment:
Unlike an ActionListener, this callback can be called multiple times. I must harden it this way or we accept a small race in the previous approach.